File: py/word_source.py
import sys
import time
import concord
from concord.computation import (
    Computation,
    Metadata,
    serve_computation
)

import logging
logging.basicConfig()
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

def time_millis():
    return int(round(time.time() * 1000))


class WordSource(Computation):
    def __init__(self):
        self.words = ['foo', 'bar', 'baz', 'fiz', 'buzz']

    def sample(self):
        """returns a random word"""
        import random
        return random.choice(self.words)

    def init(self, ctx):
        logger.info("Source initialized")
        ctx.set_timer('loop', time_millis())

    def process_timer(self, ctx, key, time):
        # stream, key, value. empty value, no need for val
        for _ in range(0, 1024):
            ctx.produce_record("words", self.sample(), '-')

        # emit records every 100ms
        ctx.set_timer("main_loop", time_millis() + 5000)

    def process_record(self, ctx, record):
        raise Exception('process_record not implemented')

    def metadata(self):
        return Metadata(
            name='word-source',
            istreams=[],
            ostreams=['words'])

logger.info("Main")
serve_computation(WordSource())

File: cpp/word_source.cc
#include <chrono>
#include <cstdint>
#include <vector>
#include <random>
#include <concord/glog_init.hpp>
#include <concord/Computation.hpp>
#include <concord/time_utils.hpp>

// sample from D. Knuth's 5 letter words dictionary
// http://www-cs-faculty.stanford.edu/~uno/sgb.html
static const std::vector<std::string> kDictionary = {
  "beach", "stock", "hurry", "saved", "sorry", "giant", "trail", "offer"};


class WordSource final : public bolt::Computation {
  public:
  using CtxPtr = bolt::Computation::CtxPtr;


  virtual void init(CtxPtr ctx) override {
    LOG(INFO) << "Initializing computation [cpp]";
    ctx->setTimer("loop", bolt::timeNowMilli());
  }

  virtual void
  processTimer(CtxPtr ctx, const std::string &key, int64_t time) override {

    const static auto kDictSize = kDictionary.size();

    for(auto i = 0u; i < 100000; ++i) {
      std::string bin = kDictionary[i % kDictSize];
      ctx->produceRecord("words", bin, "");
    }

    ctx->setTimer(key, bolt::timeNowMilli());
  }

  virtual bolt::Metadata metadata() override {
    bolt::Metadata m;
    m.name = "word-source";
    m.ostreams.insert("words");
    return m;
  }

  virtual void processRecord(CtxPtr ctx, bolt::FrameworkRecord &&r) override {
    LOG(FATAL) << "source does not impl processRecord";
  }
};

int main(int argc, char *argv[]) {
  // int i = 0; while(!i){sleep(1);} // gdb hack
  bolt::logging::glog_init(argv[0]);
  bolt::client::serveComputation(std::make_shared<WordSource>(), argc, argv);
  return 0;
}
